File: Problems/python/boj_숫자정사각형.py
#https://www.acmicpc.net/problem/1051
#Implementation

import sys
input = sys.stdin.readline

n, m = map(int, input().split())
s = min(n, m)
mat = [list(map(int, input().rstrip())) for _ in range(n)]

result = 1
for k in range(2, s+1):
    for i in range(n-k+1):
        for j in range(m-k+1):
            if mat[i][j]==mat[i][j+k-1]==mat[i+k-1][j]==mat[i+k-1][j+k-1]:
                result = max(result, k**2)
    
print(result)

File: Problems/cpp/boj_숫자정사각형.cpp
//https://www.acmicpc.net/problem/1051
//Implementation

#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int n, m;
int result=1;

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    cin >> n >> m;
    vector<string> mat(n);
    for (int i=0; i<n; i++){
        cin >> mat[i];
    }

    int len = min(n, m);

    for (int k=1; k<len; k++){
        for (int i=0; i<n-k; i++){
            for (int j=0; j<m-k; j++){
                if (mat[i][j]==mat[i][j+k] && mat[i][j+k]==mat[i+k][j] && mat[i+k][j]==mat[i+k][j+k]){
                    result = max(result, (k+1)*(k+1));
                }
            }
        }
    }

    cout << result;
}